GH 920 - Topic-based retry support (#1664)
spring-projects/spring-kafka: spring-kafka-docs/src/main/asciidoc/retrytopic.adoc

[[retry-topic]]
== Non-Blocking Retries

> 

Reviewer: I think we should add (for now).

IMPORTANT: This is an experimental feature and the usual rule of no breaking API changes does not apply to this feature until the experimental designation is removed.
Users are encouraged to try out the feature and provide feedback via GitHib Issues or GitHub discussions.

Author: Sure, I totally agree with that, done.

Achieving non-blocking retry / DLT functionality with Kafka usually requires setting up extra topics and creating and configuring the corresponding listeners.

> Since 2.7 Spring Kafka offers the `@RetryableTopic` annotation and `RetryTopicConfiguration` class to simplify that bootstraping.

Reviewer: ```suggestion
Since 2.7 Spring for Apache Kafka offers the `@RetryableTopic` annotation and `RetryTopicConfiguration` class to simplify that bootstraping.
```

The foundation requires us to mention Apache in quotes like this; we can use `spring-kafka` when referring to the jar.

Author: Done

The retry topic consumer then checks the timestamp and if it's not due it pauses the consumption for that topic's partition.
When it is due the partition consumption is resumed and the message is consumed again.
If the message processing fails again the message will be forwarded to the next retry topic, and the pattern is repeated until a successful processing occurs, or the attempts are exhausted and the message is sent to the Dead Letter Topic.

To illustrate, if you have a "main-topic" topic, and want to setup non-blocking retry with an exponential backoff of 1000ms with a multiplier of 2 and 4 max attempts, it will create the main-topic-retry-1000, main-topic-retry-2000, main-topic-retry-4000 and main-topic-dlt topics and configure the respective consumers.
The framework also takes care of creating the topics and setting up and configuring the listeners.

IMPORTANT: By using this strategy you lose Kafka's ordering guarantees for that topic.

IMPORTANT: At this time this functionality only supports ConcurrentKafkaListenerContainerFactory factories and doesn't support class level `@KafkaListener` annotations

=== Configuration

> ==== Using the `@RetriableTopic` annotation

Reviewer: ```suggestion
==== Using the `@RetryableTopic` Annotation
```

Author: Done

To configure the retry and dlt topics for a `@KafkaListener` annotated method, you just have to add the `@RetryableTopic` annotation to it and Spring Kafka will bootstrap all the necessary topics and consumers with the default configurations.

====
[source, java]
----
@RetryableTopic(kafkaTemplate = "myRetryableTopicKafkaTemplate")
@KafkaListener(topics = "my-annotated-topic")
public void processMessage(MyPojo message) {
        // ... message processing
}
----
====

You can specify a method in the same class to process the dlt messages by annotating it with the `@DltHandler` annotation.
If no DltHandler method is provided a default consumer is created which only logs the consumption.

====
[source, java]
----
@DltHandler
public void processMessage(MyPojo message) {
// ... message processing, persistence, etc
}
----
====

If you don't specify a kafkaTemplate name a bean with name `retryTopicDefaultKafkaTemplate` will be looked up.

==== Using `RetryTopicConfiguration` beans

> You can also configure the non-blocking retrial by creating `RetryTopicConfiguration` beans in a `@Configuration` annotated class.

Reviewer: ```suggestion
You can also configure the non-blocking retry support by creating `RetryTopicConfiguration` beans in a `@Configuration` annotated class.
```

Author: Done

To achieve more fine-grained control over how to handle non-blocking retrials for each topic, more than one `RetryTopicConfiguration` bean can be provided.

====
[source, java]
----
@Bean
public RetryTopicConfiguration myRetryTopic(KafkaTemplate<String, MyPojo> template) {
return RetryTopicConfiguration
            .builder()
            .fixedBackoff(3000)
            .maxAttempts(5)
            .includeTopics("my-topic", "my-other-topic")
            .create(template);
}

@Bean
public RetryTopicConfiguration myOtherRetryTopic(KafkaTemplate<String, MyOtherPojo> template) {
           return RetryTopicConfiguration
                   .builder()
                   .exponentialBackoff(1000, 2, 5000)
                   .maxAttempts(4)
                   .excludeTopics("my-topic", "my-other-topic")
                   .retryOn(MyException.class)
                   .create(template);
}
----
====

=== Features

Most of the features are available both for the `@RetryableTopic` annotation and the `RetryTopicConfiguration` beans.

==== BackOff Configuration

> The BackOff configuration relies on the `BackOff` interface from the `Spring Retry` project.

Reviewer: ```suggestion
The BackOff configuration relies on the `BackOffPolicy` interface from the `Spring Retry` project.
```

Author: Done

----
@RetryableTopic(retryTopicSuffix = "-my-retry-suffix", dltTopicSuffix = "-my-dlt-suffix")
@KafkaListener(topics = "my-annotated-topic")
public void processMessage(MyPojo message) {
        // ... message processing
}
----
====

====
[source, java]
----
@Bean
public RetryTopicConfiguration myRetryTopic(KafkaTemplate<String, MyOtherPojo> template) {
           return RetryTopicConfiguration
                   .builder()
                   .retryTopicSuffix("-my-retry-suffix")
                   .dltTopicSuffix("-my-dlt-suffix")
                   .create(template);
           }
----
====

The retry suffix also is itself suffixed with the delay for the topic, for example -1000, except for fixed delay configurations where the suffix is -0, -1, etc, and single topic fixed delay configurations which have no suffix besides the first one.

NOTE: The default suffixes are "-retry" and "-dlt", for retry and dlt topics respectively.

==== Dlt Processing

You can specify the method used to process the Dlt for the topic, as well as the behaviour if that processing fails.

To do that you can use the `@DltHandler` annotation in a method of the class with the `@RetryableTopic` annotation(s).

> Note that the same method will be used for all the `@RetryableTopic` annottated methods within that class.

Reviewer: ```suggestion
Note that the same method will be used for all the `@RetryableTopic` annotated methods within that class.
```

Author: Done

// ... message processing, persistence, etc
}
----
====

The DLT handler method can also be provided through the RetryTopicConfigurationBuilder.dltHandlerMethod(Class, String) method, providing the class and method name that should handle the DLT topic.
If a bean instance of the provided class is found in the application context that bean is used for Dlt processing, otherwise an instance is created with it's dependencies properly injected.

====
[source, java]
----
@Bean
public RetryTopicConfigurer myRetryTopic(KafkaTemplate<Integer, MyPojo> template) {
return RetryTopicConfigurer
    .builder()
    .dltProcessor(MyCustomDltProcessor.class, "processDltMessage")
    .create(template);
}

@Component
public class MyCustomDltProcessor {

    private final MyDependency myDependency;

    public MyCustomDltProcessor(MyDependency myDependency) {
        this.myDependency = myDependency;
    }

    public void processDltMessage(MyPojo message) {
       // ... message processing, persistence, etc
    }
}
----
====

NOTE: If no DLT handler is provided, the default RetryTopicConfigurer.LoggingDltListenerHandlerMethod is used.

==== Dlt Failure Behaviour

> Should the Dlt processing fail, there are two possible behaviours available: `ALWAYS_RETRY` and `ABORT`.

Reviewer: ```suggestion
Should the Dlt processing fail, there are two possible behaviors available: `ALWAYS_RETRY` and `FAIL`.
```
US English for consistency (I am from the UK, originally).
We try to avoid words like abort,

Author: Done

In the former the message is forwarded back to the dlt topic so it doesn't block other dlt messages' processing.

> In the latter the consumer aborts the execution without forwarding the message.

Reviewer: ```suggestion
In the latter the consumer ends the execution without forwarding the message.
```

Author: Done
